File: app/Models/Transaction.php
<?php

namespace App\Models;

use App\Traits\SyncsWithHQ;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Cache;

class Transaction extends Model
{
    use HasFactory, SyncsWithHQ;

    protected $fillable = [
        'account_id', 'reference_number', 'type', 'amount', 'currency',
        'balance_before', 'balance_after', 'status', 'description',
        'recipient_account_id', 'recipient_name', 'recipient_bank',
        'recipient_account_number', 'channel', 'ip_address', 'metadata',
        'completed_at',
    ];

    protected function casts(): array
    {
        return [
            'amount' => 'decimal:2',
            'balance_before' => 'decimal:2',
            'balance_after' => 'decimal:2',
            'metadata' => 'array',
            'completed_at' => 'datetime',
        ];
    }

    // ── Relationships ───────────────────────────────────────────

    public function account()
    {
        return $this->belongsTo(Account::class);
    }

    public function recipientAccount()
    {
        return $this->belongsTo(Account::class, 'recipient_account_id');
    }

    // ── Scopes ──────────────────────────────────────────────────

    public function scopeCompleted($query)
    {
        return $query->where('status', 'completed');
    }

    public function scopeOfType($query, string $type)
    {
        return $query->where('type', $type);
    }

    public function scopeRecent($query, int $days = 30)
    {
        return $query->where('created_at', '>=', now()->subDays($days));
    }

    // ── Helpers ─────────────────────────────────────────────────

    public function isCredit(): bool
    {
        return in_array($this->type, ['deposit', 'transfer_in', 'refund', 'interest', 'loan_disbursement']);
    }

    public function isDebit(): bool
    {
        return in_array($this->type, ['withdrawal', 'transfer_out', 'payment', 'fee', 'loan_repayment']);
    }

    public function getCurrencyModel(): ?Currency
    {
        return Cache::remember("currency_{$this->currency}", 3600, function () {
            return Currency::where('code', $this->currency)->first();
        });
    }

    public function getFormattedAmountAttribute(): string
    {
        $prefix = $this->isCredit() ? '+' : '-';
        $currency = $this->getCurrencyModel();
        if ($currency) {
            return $prefix . $currency->symbol . ' ' . number_format($this->amount, $currency->decimal_places);
        }
        return $prefix . $this->currency . ' ' . number_format($this->amount, 2);
    }

    public function getStatusColorAttribute(): string
    {
        return match($this->status) {
            'completed' => 'green',
            'pending' => 'yellow',
            'failed' => 'red',
            'cancelled' => 'gray',
            'reversed' => 'orange',
            default => 'gray',
        };
    }

    public function getTypeIconAttribute(): string
    {
        return match($this->type) {
            'deposit' => 'arrow-down-circle',
            'withdrawal' => 'arrow-up-circle',
            'transfer_in' => 'arrow-down-left',
            'transfer_out' => 'arrow-up-right',
            'payment' => 'shopping-cart',
            'refund' => 'refresh-cw',
            'fee' => 'percent',
            'interest' => 'trending-up',
            'loan_disbursement' => 'dollar-sign',
            'loan_repayment' => 'credit-card',
            default => 'activity',
        };
    }

    public static function generateReference(): string
    {
        return 'TXN' . strtoupper(bin2hex(random_bytes(10)));
    }
}

File: mobile/lib/data/models/transaction.dart
class Transaction {
  final int id;
  final int accountId;
  final String referenceNumber;
  final String type;
  final double amount;
  final String currency;
  final double? balanceBefore;
  final double? balanceAfter;
  final String status;
  final String? description;
  final String? recipientName;
  final String? channel;
  final String? completedAt;
  final String? createdAt;

  const Transaction({
    required this.id,
    required this.accountId,
    required this.referenceNumber,
    required this.type,
    required this.amount,
    required this.currency,
    this.balanceBefore,
    this.balanceAfter,
    required this.status,
    this.description,
    this.recipientName,
    this.channel,
    this.completedAt,
    this.createdAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as int,
      accountId: json['account_id'] as int,
      referenceNumber: json['reference_number'] as String,
      type: json['type'] as String,
      amount: (json['amount'] as num).toDouble(),
      currency: json['currency'] as String,
      balanceBefore: json['balance_before'] != null
          ? (json['balance_before'] as num).toDouble()
          : null,
      balanceAfter: json['balance_after'] != null
          ? (json['balance_after'] as num).toDouble()
          : null,
      status: json['status'] as String,
      description: json['description'] as String?,
      recipientName: json['recipient_name'] as String?,
      channel: json['channel'] as String?,
      completedAt: json['completed_at'] as String?,
      createdAt: json['created_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'account_id': accountId,
      'reference_number': referenceNumber,
      'type': type,
      'amount': amount,
      'currency': currency,
      'balance_before': balanceBefore,
      'balance_after': balanceAfter,
      'status': status,
      'description': description,
      'recipient_name': recipientName,
      'channel': channel,
      'completed_at': completedAt,
      'created_at': createdAt,
    };
  }

  /// Whether this transaction is a credit (incoming money).
  bool get isCredit =>
      type == 'deposit' || type == 'refund' || type == 'interest';
}
